### PR TITLE
Add weighted song selection shuffle types

### PR DESCRIPTION
**Pull request by Brainicism**

### CODE REVIEW

File: src/types.ts [hunk 1]
```diff
@@ -41,6 +41,7 @@ export interface QueriedSong {
     tags?: string;
     language?: string;
     vtype?: string;
+    selectionWeight?: number;
```
Comment (taahamahdi) [inline src/types.ts:44]: Can this be made required?

**Review by taahamahdi — DISMISSED** [review]

File: src/types.ts (hunk 1 above)
Comment (Brainicism) [reply src/types.ts:44]: I don't know why any of these are nullable tbh

File: src/commands/game_options/shuffle.ts [hunk 2]
```diff
@@ -86,7 +95,18 @@ export default class ShuffleCommand implements BaseCommand {
         const shuffleType =
             parsedMessage.components[0].toLowerCase() as ShuffleType;
 
-        guildPreference.setShuffleType(shuffleType);
+        if (PREMIUM_SHUFFLE_TYPES.includes(shuffleType)) {
+            if (!(await isUserPremium(message.author.id))) {
+                sendErrorMessage(MessageContext.fromMessage(message), {
+                    description:
+                        "This option can only be used by premium KMQ supporters.",
+                    title: "Premium Option",
```
Comment (taahamahdi) [inline src/commands/game_options/shuffle.ts:103]: Should move these to i18n

**Review by taahamahdi — CHANGES_REQUESTED** [review]

File: src/commands/game_options/special.ts [hunk 3]
```diff
@@ -172,11 +149,6 @@ export default class SpecialCommand implements BaseCommand {
     call = async ({ message, parsedMessage }: CommandArgs): Promise<void> => {
         const guildPreference = await getGuildPreference(message.guildID);
         if (parsedMessage.components.length === 0) {
-            await resetSpecial(
```
Comment (taahamahdi) [inline src/commands/game_options/special.ts:175]: Missing a reset here?

File: src/structures/game_session.ts [hunk 4]
```diff
@@ -705,19 +704,20 @@ export default class GameSession extends Session {
      * The game has changed its premium state, so update filtered songs/remove ,special
      */
     async updatePremiumStatus(): Promise<void> {
-        await this.reloadSongs(await getGuildPreference(this.guildID));
-
         const guildPreference = await getGuildPreference(this.guildID);
-        if (
-            !this.isPremiumGame() &&
-            this.guildID !== process.env.DEBUG_SERVER_ID &&
```
Comment (taahamahdi) [inline src/structures/game_session.ts:713]: We probably want to let users keep using special in the kmq server

**Pull request merged**